xla/generate/base.py

```diff
@@ -0,0 +1,187 @@
+import json
+import sys
+import time
+from pathlib import Path
+from typing import Optional
+
+import lightning as L
+import torch
+from lightning.fabric.strategies import XLAFSDPStrategy
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.parent.resolve()
+sys.path.append(str(wd))
+
+import torch_xla.core.xla_model as xm
+
+from lit_gpt import GPT, Config, Tokenizer
+from lit_gpt.utils import check_valid_checkpoint_dir, lazy_load
+
+
+@torch.no_grad()
+def generate(
+    model: torch.nn.Module,
+    idx: torch.Tensor,
+    max_returned_tokens: int,
+    max_seq_length: int,
+    *,
+    temperature: float = 1.0,
+    top_k: Optional[int] = None,
+    eos_id: Optional[int] = None,
+) -> torch.Tensor:
+    """Takes a conditioning sequence (prompt) as input and continues to generate as many tokens as requested.
+
+    The implementation of this function is modified from A. Karpathy's nanoGPT.
+
+    Args:
+        model: The model to use.
+        idx: Tensor of shape (T) with indices of the prompt sequence.
+        max_returned_tokens: The maximum number of tokens to return (given plus generated).
+        max_seq_length: The maximum sequence length allowed. Should be less or equal than the block size.
+        temperature: Scales the predicted logits by 1 / temperature.
+        top_k: If specified, only sample among the tokens with the k highest probabilities.
+        eos_id: If specified, stop generating any more token once the <eos> token is triggered.
+    """
+    T = idx.size(0)
+    assert max_returned_tokens > T
+    device, dtype = idx.device, idx.dtype
+    # create an empty tensor of the expected final shape and fill in the current tokens
+    empty = torch.empty(max_returned_tokens, dtype=dtype, device=device)
+    empty[:T] = idx
+    idx = empty
+    input_pos = torch.arange(0, T, device=device)
+
+    xm.mark_step()
+
+    # generate up to a fixed number of tokens
+    for _ in range(max_returned_tokens - T):
+        x = idx.index_select(0, input_pos).view(1, -1)
+
+        # forward
+        logits = model(x, max_seq_length, input_pos)
```

Comment:
This currently has a bug on nightly where https://github.com/Lightning-AI/lit-gpt/blob/main/lit_gpt/model.py#L82 triggers
```python
2023-08-24 00:06:20.680726: F ./torch_xla/csrc/runtime/debug_macros.h:20] Non-OK-status: status.status() status: INVALID_ARGUMENT: Input dimension should be either 1 or equal to the output dimension it is broadcasting into; the 0th operand dimension is 7, the 0th output dimension is 1.
*** Begin stack trace ***
    tsl::CurrentStackTrace[abi:cxx11]()
    xla::Shape const* ConsumeValue<xla::Shape const*>(tsl::StatusOr<xla::Shape const*>&&)
    torch_xla::ShapeHelper::ShapeOfXlaOp(xla::XlaOp)
    torch_xla::InferOutputShape(absl::lts_20230125::Span<xla::Shape const>, std::function<xla::XlaOp (absl::lts_20230125::Span<xla::XlaOp const>)> const&)
    
    
    torch_xla::XlaNode::GetOpShape(std::function<xla::Shape ()> const&) const
    torch_xla::XlaNode::XlaNode(torch::lazy::OpKind, c10::ArrayRef<torch::lazy::Value>, std::function<xla::Shape ()> const&, unsigned long, torch::lazy::hash_t)
    torch_xla::Expand::Expand(torch::lazy::Value const&, std::vector<long, std::allocator<long> >)
    torch_xla::EnsureRank1(torch::lazy::Value const&)
    torch_xla::tensor_methods::index_select(c10::intrusive_ptr<torch_xla::XLATensor, c10::detail::intrusive_target_default_null_type<torch_xla::XLATensor> > const&, long, c10::intrusive_ptr<torch_xla::XLATensor, c10::detail::intrusive_target_default_null_type<torch_xla::XLATensor> > const&)
    torch_xla::XLANativeFunctions::index_select(at::Tensor const&, long, at::Tensor const&)
    
    c10::Dispatcher::callBoxed(c10::OperatorHandle const&, std::vector<c10::IValue, std::allocator<c10::IValue> >*) const
    
    at::_ops::index_select::redispatch(c10::DispatchKeySet, at::Tensor const&, long, at::Tensor const&)
    
    
    at::_ops::index_select::call(at::Tensor const&, long, at::Tensor const&)
```


Strangely, doing `mark_step()` before `forward` doesn't fix it, but doing it at the very beginning of `forward` does.

If there's no fix for this, we could work around it by making `input_pos` be always length 1

cc @gkroiz

Reply:
This behavior seems quite bizarre and when we originally added xla support several months back this bug did not exist. 

```
INVALID_ARGUMENT: Input dimension should be either 1 or equal to the output dimension it is broadcasting into; the 0th operand dimension is 7, the 0th output dimension is 1.
```
The above suggests that this issue is related to `input_pos`, however, `input_pos` is defined before the first forward pass and there is a `mark_step()` in between the `input_pos` definition and the first forward pass. In the forward pass, `input_pos` is not updated.

cc @liyang90 do you have an idea as to why this is happening?

Reply:
It happens with FSDP only (not ordinary spawning strategy). I'll use this workaround in the meantime: https://github.com/Lightning-AI/lit-gpt/pull/435/commits/4b06f540c987a1e154b0de4cd0f4f80832503dd9 but you might want to debug this further and fix it since it's most likely a bug in the XLA FSDP implementation